### ISSUE TITLE
关于sampled softmax loss

### ISSUE THREAD

**Issue by xbingsun** [bbruceyuan/DeepMatch-Torch#3]
您好，MIND模型中提到计算loss时候使用了sampled softmax的方法，在deepmatch的代码里我看到有对tf的版本调用，但是在pytorch版本里面好像只传入了sample的个数，但是并没有使用，请问这部分是否会考虑实现？

**Comment by xbingsun** [comment]
我刚刚有看到代码里写的“sample softmax 可以通过 构造样本实现”，请问应该怎么构造呢？麻烦啦！

**Comment by bbruceyuan** [comment]
在 PyTorch 里面，没有 sampledsoftmax_loss 函数，所以我通过 自己构造样本 [1, 0, 0, 0, 0] 进行 CrossEntropyLoss 来代替 sampled_softmax_loss, 1是正样本， 0 是采样的, 让 第一个元素用于是 正样本，这样进行交叉熵可以起到类似 sampledsoftmax 的作用。

那么在 PyTorch 里面的 [CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) 的输入参数是 Target, shape 是 (N)；因为我们用交叉熵做 loss, 所以 label 是第一个位置，因此 label 是 0。

代码部分可以查看：https://github.com/bbruceyuan/DeepMatch-Torch/blob/322589cbe7f6d4efb1c4094bd6fb62f599f62c29/deepmatch_torch/models/PLBaseModel.py 中 _step 函数；



如果有其他疑问可以直接回复~

其他类似的问题：https://github.com/bbruceyuan/DeepMatch-Torch/discussions/1

**Comment by xbingsun** [comment]
明白啦～谢谢你的回复，我还有一个小问题是在数据处理的时候我看到deepmatch和pytorch版本都将用户序列倒序了（ hist[::-1]），请问这个操作的目的是什么呢？前面已经通过data.sort_values("timestamp", inplace=True) 对用户行为升序排序之后，是否应该保留升序的顺序？
参考代码：`train_set.append((reviewerID, hist[::-1], pos_list[i], 1, len(hist[::-1]),rating_list[i]))`

**Comment by bbruceyuan** [comment]
这个我也是直接抄的 DeepMatch实现，按照我的理解这个应该是【倒序与否】无所谓的。反正都是 pooling 得到 embedding。

> 明白啦～谢谢你的回复，我还有一个小问题是在数据处理的时候我看到deepmatch和pytorch版本都将用户序列倒序了（ hist[::-1]），请问这个操作的目的是什么呢？前面已经通过data.sort_values("timestamp", inplace=True) 对用户行为升序排序之后，是否应该保留升序的顺序？ 参考代码：`train_set.append((reviewerID, hist[::-1], pos_list[i], 1, len(hist[::-1]),rating_list[i]))`

**Comment by xbingsun** [comment]
嗯呢！对于mind这样不涉及时序关系的模型来说应该没有什么问题，只是如果在例如sdm这样的模型里面可能会有影响
